File: solutions/exercise-4/src/main/java/io/miragon/training/domain/Email.java
package io.miragon.training.domain;

public record Email(String value) {

    public Email {
        if (!value.matches("^[^@\\s]+@[^@\\s]+\\.[^@\\s]+$")) {
            throw new IllegalArgumentException("Invalid email: " + value);
        }
    }
}

File: solutions/exercise-6/src/main/kotlin/io/miragon/training/domain/Email.kt
package io.miragon.training.domain

data class Email(val value: String) {
    init {
        require(value.matches(Regex("^[^@\\s]+@[^@\\s]+\\.[^@\\s]+$"))) {
            "Invalid email address: $value"
        }
    }
}
